sdk/nodejs/appservice/zMixins.ts

```diff
@@ -334,87 +432,125 @@ export class CallbackFunctionApp<C extends Context<R>, E, R extends Result> exte
+            }
+        }
 
         let plan = args.plan;
         if (!plan) {
             plan = new appservice.Plan(name, {
-                ...resourceGroupArgs,
+                resourceGroupName: args.resourceGroupName,
 
                 kind: "FunctionApp",
 
                 sku: {
                     tier: "Dynamic",
                     size: "Y1",
                 },
-            }, opts);
+            }, { parent: this });
         }
 
         const storageMod = <typeof storageForTypesOnly>require("../storage");
         const account = args.account || new storageMod.Account(makeSafeStorageAccountName(name), {
-            ...resourceGroupArgs,
+            resourceGroupName: args.resourceGroupName,
 
             accountKind: "StorageV2",
             accountTier: "Standard",
             accountReplicationType: "LRS",
-        }, opts);
+        }, { parent: this });
 
         const container = args.container || new storageMod.Container(makeSafeStorageContainerName(name), {
             resourceGroupName: args.resourceGroupName,
             storageAccountName: account.name,
             containerAccessType: "private",
-        }, opts);
+        }, { parent: this });
+
+        const archive = args.archive !== undefined 
+            ? args.archive
+            : pulumi.output(produceDeploymentPackage(args)).apply(m => new pulumi.asset.AssetArchive(m));
 
-        const assetMap = serializeCallback(name, args, bindings);
         const zipBlob = new storageMod.ZipBlob(name, {
             resourceGroupName: args.resourceGroupName,
             storageAccountName: account.name,
             storageContainerName: container.name,
             type: "block",
-            content: assetMap.apply(m => new pulumi.asset.AssetArchive(m)),
-        }, opts);
+            content: archive,
+        }, { parent: this });
 
         const codeBlobUrl = storageMod.signedBlobReadUrl(zipBlob, account);
-        super(name, {
+        this.app = new FunctionApp(name, {
             ...args,
-            ...resourceGroupArgs,
 
-            version: "~2",
             appServicePlanId: plan.id,
             storageConnectionString: account.primaryConnectionString,
+            version: args.version || "~2",
 
-            appSettings: pulumi.output(args.appSettings).apply(settings => {
-                settings = settings || {};
+            appSettings: combineAppSettings(args).apply(settings => {
                 return {
                     ...settings,
-                    WEBSITE_RUN_FROM_ZIP: codeBlobUrl,
+                    WEBSITE_RUN_FROM_PACKAGE: codeBlobUrl,
                     WEBSITE_NODE_DEFAULT_VERSION: util.ifUndefined(args.nodeVersion, "8.11.1"),
                 };
             }),
-        }, opts);
+        }, { parent: this });
 
         this.account = account;
         this.container = container;
         this.plan = plan;
         this.zipBlob = zipBlob;
+        this.functions = args.functions;

```

Comment:
should we expose hte optional archive for consistency here?

Reply:
I actually made functions private, thinking to use them in `getEndpoint`, but ended up not using them. I think it might be confusing to expose them and then not filling them in case of the archive scenario. Removed for now.